### PR TITLE
Annotations

### PR DESCRIPTION
**Pull request by TomBrouws**
Parsing of gene annotation data is now implemented, along with incorporating this data in our graph data structure. You can also add new labels to nodes next to the labels that are already placed during parsing.

### CODE REVIEW

**Comment by Delfil** [comment]
I think we want the labels in a dedicated package.
tudelft.ti2806.pl3.data.label?

File: src/main/java/tudelft/ti2806/pl3/data/graph/EndGeneLabel.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+			return false;
+		}
+		EndGeneLabel that = (EndGeneLabel) o;
+
+		return endRef == that.endRef;
+
+	}
+
+	@Override
+	public int hashCode() {
```
Comment (Delfil) [inline src/main/java/tudelft/ti2806/pl3/data/graph/EndGeneLabel.java]: Should include text field of superclass.

File: src/main/java/tudelft/ti2806/pl3/data/graph/StartGeneLabel.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+			return false;
+		}
+		StartGeneLabel that = (StartGeneLabel) o;
+
+		return startRef == that.startRef;
+
+	}
+
+	@Override
+	public int hashCode() {
```
Comment (Delfil) [inline src/main/java/tudelft/ti2806/pl3/data/graph/StartGeneLabel.java]: Should include text field of superclass.

File: src/main/java/tudelft/ti2806/pl3/data/graph/StartGeneLabel.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Created by tombrouws on 27/05/15.
+ */
+public class StartGeneLabel extends Label {
+
+	private final int startRef;
+
+	@Override
+	public boolean equals(Object o) {
```
Comment (Delfil) [inline src/main/java/tudelft/ti2806/pl3/data/graph/StartGeneLabel.java]: Should include text field of superclass.

File: src/main/java/tudelft/ti2806/pl3/data/graph/Label.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	public Label(String s) {
+		if (s != null) {
+			text = s;
+		} else {
+			text = "";
+		}
+	}
+
+	@Override
+	public boolean equals(Object o) {
```
Comment (Delfil) [inline src/main/java/tudelft/ti2806/pl3/data/graph/Label.java]: Should include text field of superclass.

File: src/main/java/tudelft/ti2806/pl3/data/graph/GraphDataRepository.java [hunk 5]
```diff
@@ -99,6 +104,35 @@ public static GraphDataRepository parseGraph(File nodesFile, File edgesFile)
...
+	 * 				the gene annotation dataset
+	 */
+	protected static void addRefLabels(DataNode node, GeneData geneData) {
+		int start = node.getRefStartPoint();
+		int end = node.getRefEndPoint();
+		Gene g = null;
+
+		boolean started = false;
+		for (int i = start; i <= end; i++) {
+			if (started) {
```
Comment (Delfil) [inline src/main/java/tudelft/ti2806/pl3/data/graph/GraphDataRepository.java]: It is possible for a label to start on multiple nodes.

**Comment by Delfil** [comment]
There is no visual difference, don't see any annotations on the graph, and I can't find any code which is supposed to add it to the graph. If it doesn't add any functionality yet or fix anything, why should we pull it?

**Comment by TomBrouws** [comment]
a. A pull request doesn't have to add any visual functionality, a backend structure is also fine. This is implemented.
b. This is so Mathieu and I can work on displaying this info more easily :)
c. Will look at the rest

**Comment by Delfil** [comment]
Kay, fine then : )

File: src/main/java/tudelft/ti2806/pl3/data/graph/GraphDataRepository.java (hunk 5 above)
Comment (TomBrouws) [inline src/main/java/tudelft/ti2806/pl3/data/graph/GraphDataRepository.java]: Discussed this together and concluded this is not a problem.

**Comment by TomBrouws** [comment]
All is fixed & more!

File: src/main/java/tudelft/ti2806/pl3/data/graph/DataNode.java [hunk 6]
```diff
@@ -34,7 +38,7 @@
 	 *            the size of this {@code Node}
 	 */
 	public DataNode(int nodeId, Genome[] source, int refStartPoint,
```
Comment (borismattijssen) [inline src/main/java/tudelft/ti2806/pl3/data/graph/DataNode.java]: Id be nice to also have a constructor without the labelList parameter and then call this constructor with the null value.

Comment (TomBrouws) [inline src/main/java/tudelft/ti2806/pl3/data/graph/DataNode.java]: thx will do

**Comment by kwendel** [comment]
Looks good!

**Comment by borismattijssen** [comment]
Like (Y)

**Pull request merged**